Kernel/FileSystem/ISO9660FileSystem.h

```diff
@@ -0,0 +1,395 @@
+
+    u8 file_structure_version; // Always 0x01.
+    u8 unused4;
+    u8 application_use[512];
+    u8 reserved[653];
+};
+static_assert(sizeof(PrimaryVolumeDescriptor) == 2048);
+
+// 8.6 Volume Partition Descriptor
+struct [[gnu::packed]] VolumePartitionDescriptor {
+    VolumeDescriptorHeader header;
+    u8 unused;
+
+    u8 system_identifier[32];
+    u8 volume_partition_identifier[32];
+    LittleAndBigEndian<u32> volume_partition_location;
+    LittleAndBigEndian<u32> volume_partition_size;
+
+    u8 system_use[1960];
+};
+static_assert(sizeof(VolumePartitionDescriptor) == 2048);
+
+}
+
+enum class RecursionDecision {
+    Recurse,
+    Continue,
+    Break
+};
+
+class ISO9660Inode;
+class ISO9660DirectoryIterator;
+
+class ISO9660FS final : public BlockBasedFileSystem {
+    friend ISO9660Inode;
+    friend ISO9660DirectoryIterator;
+
+public:
+    struct DirectoryEntry : public RefCounted<DirectoryEntry> {
+        u32 extent { 0 };
+        u32 length { 0 };
+
+        // NOTE: This can never be empty if we read the directory successfully.
+        //       We need it as an OwnPtr to default-construct this struct.
+        OwnPtr<KBuffer> blocks;
+
+        static KResultOr<NonnullRefPtr<DirectoryEntry>> create(u32 extent, u32 length, OwnPtr<KBuffer> blocks)
+        {
+            auto result = adopt_ref_if_nonnull(new (nothrow) DirectoryEntry(extent, length, move(blocks)));
+            if (!result) {
+                return ENOMEM;
+            }
+            return result.release_nonnull();
+        }
+
+    private:
+        DirectoryEntry(u32 extent, u32 length, OwnPtr<KBuffer> blocks)
+            : extent(extent)
+            , length(length)
+            , blocks(move(blocks))
+        {
+        }
+    };
+
+    static KResultOr<NonnullRefPtr<ISO9660FS>> create(FileDescription&);
+
+    virtual ~ISO9660FS() override;
+    virtual bool initialize() override;
+    virtual StringView class_name() const override { return "ISO9660FS"; }
+    virtual Inode& root_inode() override;
+
+    virtual unsigned total_block_count() const override;
+    virtual unsigned total_inode_count() const override;
+
+    virtual u8 internal_file_type_to_directory_entry_type(DirectoryEntryView const& entry) const override;
+
+    KResultOr<NonnullRefPtr<DirectoryEntry>> directory_entry_for_record(Badge<ISO9660DirectoryIterator>, ISO::DirectoryRecordHeader const* record);
+
+private:
+    ISO9660FS(FileDescription&);
+
+    bool parse_volume_set();
+    bool create_root_inode();
+    bool calculate_inode_count() const;
+
+    u32 calculate_directory_entry_cache_key(ISO::DirectoryRecordHeader const*);
+
+    KResult visit_directory_record(ISO::DirectoryRecordHeader const* record, Function<KResultOr<RecursionDecision>(ISO::DirectoryRecordHeader const*)> const& visitor) const;
+
+    OwnPtr<ISO::PrimaryVolumeDescriptor> m_primary_volume;
+    RefPtr<ISO9660Inode> m_root_inode;
+
+    mutable unsigned m_cached_inode_count { 0 };
```

Comment:
`unsigned` seems like an arbitrary type here. Either use `size_t`, or if a specific bit width is desired, use e.g `u16` / `u32` / `u64`

Reply:
Done, but we should also convert `unsigned`s in `total_{block,inode}_count` to an explicitly sized value then.